File: linux/Lockroot.Linux/Security/KdfParams.cs
using System.Security.Cryptography;

namespace Lockroot.Linux.Security;

public sealed class KdfParams
{
    public const int MinMemoryKiB = 19_456;
    public const int MaxMemoryKiB = 262_144;
    public const int MinIterations = 2;
    public const int MaxIterations = 10;
    public const int MinParallelism = 1;
    public const int MaxParallelism = 8;
    public const int MinSaltBytes = 16;
    public const int MaxSaltBytes = 64;

    public int MemoryKiB { get; set; }
    public int Iterations { get; set; }
    public int Parallelism { get; set; }
    public byte[] Salt { get; set; } = [];

    public void Validate()
    {
        if (MemoryKiB is < MinMemoryKiB or > MaxMemoryKiB)
        {
            throw new CryptographicException("Unsupported Argon2id memory parameter.");
        }

        if (Iterations is < MinIterations or > MaxIterations)
        {
            throw new CryptographicException("Unsupported Argon2id iteration parameter.");
        }

        if (Parallelism is < MinParallelism or > MaxParallelism)
        {
            throw new CryptographicException("Unsupported Argon2id parallelism parameter.");
        }

        if (Salt.Length is < MinSaltBytes or > MaxSaltBytes)
        {
            throw new CryptographicException("Unsupported Argon2id salt length.");
        }
    }
}

File: macos/LockrootMac/Sources/LockrootMac/Crypto/KdfParams.swift
import Foundation

struct KdfParams: Equatable {
    static let minMemoryKiB = 19_456
    static let maxMemoryKiB = 262_144
    static let minIterations = 2
    static let maxIterations = 10
    static let minParallelism = 1
    static let maxParallelism = 8
    static let minSaltBytes = 16
    static let maxSaltBytes = 64

    let memoryKiB: Int
    let iterations: Int
    let parallelism: Int
    let salt: Data

    init(memoryKiB: Int, iterations: Int, parallelism: Int, salt: Data) throws {
        guard (Self.minMemoryKiB...Self.maxMemoryKiB).contains(memoryKiB) else {
            throw CryptoError.unsupportedFormat("Unsupported Argon2id memory cost.")
        }
        guard (Self.minIterations...Self.maxIterations).contains(iterations) else {
            throw CryptoError.unsupportedFormat("Unsupported Argon2id iteration count.")
        }
        guard (Self.minParallelism...Self.maxParallelism).contains(parallelism) else {
            throw CryptoError.unsupportedFormat("Unsupported Argon2id parallelism.")
        }
        guard (Self.minSaltBytes...Self.maxSaltBytes).contains(salt.count) else {
            throw CryptoError.unsupportedFormat("Unsupported Argon2id salt length.")
        }

        self.memoryKiB = memoryKiB
        self.iterations = iterations
        self.parallelism = parallelism
        self.salt = salt
    }
}
